src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,182 @@
+            argument_types_, params)
+        , threshold(threshold_)
+        , input_data_type(this->argument_types[0])
+    {
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>::create(
+            place);
+        this->data(place).threshold = threshold;
+    }
+
+    String getName() const override { return is_weighted ? "groupSortedArrayWeighted" : "groupSortedArray"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeArray>(input_data_type); }
+
+    bool allocatesMemoryInArena() const override
+    {
+        if constexpr (std::is_same_v<T, StringRef>)
+            return true;
+        else
+            return false;
+    }
+
+    void add(AggregateDataPtr __restrict, const IColumn **, size_t, Arena *) const override { }
+
+    void addBatchSinglePlace(
+        size_t batch_size, AggregateDataPtr place, const IColumn ** columns, Arena * arena, ssize_t if_argument_pos) const override
+    {
+        State stateAux(threshold);
+
+        if constexpr (is_weighted)
+        {
+            //First store the first n elements with the column number
+            AggregateFunctionGroupSortedArrayData<size_t, true> mapAux(threshold);
+            if (if_argument_pos >= 0)
+            {
+                const auto & flags = assert_cast<const ColumnUInt8 &>(*columns[if_argument_pos]).getData();
+                for (size_t i = 0; i < batch_size; ++i)
+                {
+                    if (flags[i])
+                        mapAux.add(i, readItem<size_t, is_plain_column>(columns[1], arena, i));
+                }
+            }
+            else
+            {
+                for (size_t i = 0; i < batch_size; ++i)
+                    mapAux.add(i, readItem<size_t, is_plain_column>(columns[1], arena, i));
+            }
+
+            //Now create a new map with the final type extracting values from selected columns
+            for (auto item : mapAux.values)
+                stateAux.add(readItem<T, is_plain_column>(columns[0], arena, item.second), item.first);
+        }
+        else
+        {
+            if (if_argument_pos >= 0)
+            {
+                const auto & flags = assert_cast<const ColumnUInt8 &>(*columns[if_argument_pos]).getData();
+                for (size_t i = 0; i < batch_size; ++i)
+                    if (flags[i])
+                    {
+                        auto val = readItem<T, is_plain_column>(columns[0], arena, i);
+                        stateAux.add(val);
+                    }
+            }
+            else
+            {
+                for (size_t i = 0; i < batch_size; ++i)
+                    stateAux.add(readItem<T, is_plain_column>(columns[0], arena, i));
+            }
+        }
+
+        State & data = this->data(place);
```

Comment:
Why is this mutex needed? Is it possible to have `add` or `addBatch...` to the same object in parallel? I don't recall seeing anything like this anywhere else.